lib/Sema/CSDiag.cpp

```diff
@@ -1030,15 +1024,16 @@ static bool findGenericSubstitutions(DeclContext *dc, Type paramType, Type actua
       if (dc && original->isTypeParameter())
         original = ArchetypeBuilder::mapTypeIntoContext(dc, original);
       
-      if (auto archetype = original->getAs<ArchetypeType>()) {
-        archetypes.push_back(archetype);
-        substitutions.push_back(substitution->getReplacementType());
-      }
+      Type replacement = substitution->getReplacementType();
+      if (auto archetype = original->getAs<ArchetypeType>())
+        archetypesMap[archetype] = replacement;
+      if (auto reverse = replacement->getAs<ArchetypeType>())
+        archetypesMap[reverse] = original;
```

Comment:
Yeah, it makes sense. I guess the ApplyExpr hasn't been built yet -- the constraint system failed, so we cannot build it.

What about this. You start with the original interface type of isEqual, which is

    <Self : EqualComparable> Self -> Self -> ()

When you substitute in the type of 't1', you check that the archetype 'T' conforms to EqualComparable, and you get T -> T -> (). Then you consider the second argument, 'u'. At this point, the original parameter type is not a generic parameter, it already has an archetype. So you can check for equality. Since its not equal to U, there's your mismatch.